cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh

```diff
@@ -0,0 +1,961 @@
+    }
+
+    const size_t idxTileBase = idxTile * size_t(tile_size);
+    _CCCL_ASSERT(idxTileBase < params.numElem, "");
+    const int valid_items   = static_cast<int>(cuda::std::min(params.numElem - idxTileBase, size_t(tile_size)));
+    const bool is_last_tile = valid_items < tile_size;
+    CpAsyncOobInfo loadInfo = prepareCpAsyncOob(params.ptrIn + idxTileBase, valid_items);
+
+    if (squad == squadLoad)
+    {
+      ////////////////////////////////////////////////////////////////////////////////
+      // Load current tile
+      ////////////////////////////////////////////////////////////////////////////////
+      SmemRef refInOutW = phaseInOutW.acquireRef();
+      squadLoadBulk(squad, refInOutW, loadInfo);
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    // Get next tile index from shared memory (all squads)
+    ////////////////////////////////////////////////////////////////////////////////
+    uint4 regNextBlockIdx{};
+    {
+      SmemRef refNextBlockIdxR = phaseNextBlockIdxR.acquireRef();
+      regNextBlockIdx          = refNextBlockIdxR.data();
+      refNextBlockIdxR.setFenceLdsToAsyncProxy();
+    }
+    bool nextIdxTileValid = ::cuda::ptx::clusterlaunchcontrol_query_cancel_is_canceled(regNextBlockIdx);
+
+    if (squad == squadReduce)
+    {
+      const int valid_items_this_thread =
+        cuda::std::clamp(valid_items - squad.threadRank() * elemPerThread, 0, elemPerThread);
+      const int valid_threads_this_warp =
+        cuda::std::clamp(::cuda::ceil_div(valid_items, elemPerThread) - squad.warpRank() * 32, 0, 32);
+      const int valid_warps = ::cuda::ceil_div(valid_items, elemPerThread * 32);
+      _CCCL_ASSERT(0 < valid_warps && valid_warps <= squad.warpCount(), "");
+
+      ////////////////////////////////////////////////////////////////////////////////
+      // Load tile from shared memory
+      ////////////////////////////////////////////////////////////////////////////////
+      AccumT regThreadSum;
+      AccumT regWarpSum;
+      {
+        // Acquire phaseInOutRW in this short scope
+        SmemRef refInOutRW = phaseInOutRW.acquireRef();
+        // Load data
+        AccumT regInput[elemPerThread];
+        // Handle unaligned refInOutRW.data() + loadInfo.smemStartOffsetElem points to the first element of the tile.
+        // in the last tile, we load some invalid elements, but don't process them later
+        squadLoadSmem(squad, regInput, &refInOutRW.data()[0] + loadInfo.smemStartOffsetElem);
+
+        ////////////////////////////////////////////////////////////////////////////////
+        // Reduce across thread and warp
+        ////////////////////////////////////////////////////////////////////////////////
+        if (is_last_tile)
+        {
+          regThreadSum = ThreadReducePartial(regInput, scan_op, valid_items_this_thread);
+          regWarpSum   = warpReducePartial(regThreadSum, scan_op, valid_threads_this_warp);
```

Comment:
TODO: for operators where we know the identity, we can use at least a better warp reduce.